### PR TITLE
Add FP8 ONNX tests

### PR DESCRIPTION
**Pull request by ahsan-ca**
This PR adds ONNX tests for FP8.

### CODE REVIEW

File: test/onnx/parse/shrink_fp8_test.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    auto cond1   = add_common_op(*mm, migraphx::make_op("less"), {x, lit_neg_lambd});
+    auto cond2_a = add_common_op(*mm, migraphx::make_op("not"), {cond1});
+    auto cond2_b = add_common_op(*mm, migraphx::make_op("greater"), {x, lit_lambd});
+    auto cond2   = add_common_op(*mm, migraphx::make_op("logical_and"), {cond2_a, cond2_b});
+
+    auto mul1 = mm->add_instruction(
+        migraphx::make_op("convert", {{"target_type", migraphx::shape::fp8e4m3fnuz_type}}), cond1);
+    auto mul2 = mm->add_instruction(
+        migraphx::make_op("convert", {{"target_type", migraphx::shape::fp8e4m3fnuz_type}}), cond2);
```
Comment (umangyadav) [inline test/onnx/parse/shrink_fp8_test.cpp:50]: `add_common_op` doesn't work for the Fp8 ? `add_common_op` should insert necessary converts   .

File: test/onnx/verify/add_fp8_test.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+TEST_CASE(add_fp8_test)
+{
+    auto p = migraphx::parse_onnx("add_fp8_test.onnx");
+    p.compile(migraphx::make_target("ref"));
+
+    migraphx::shape s{migraphx::shape::fp8e4m3fnuz_type, {1}};
+
+    migraphx::parameter_map pp;
+    migraphx::literal l1{s, {3.0}};
+    migraphx::literal l2{s, {2.0}};
```
Comment (umangyadav) [inline test/onnx/verify/add_fp8_test.cpp:39]: I find it better to use floating point numbers with fractional part for the test. 

for example,  you can pick any numbers from the FP8 list for the test. 
https://github.com/ROCm/AMDMIGraphX/blob/47b5f10772183b7f16aa4f3a173ae23c525c23ae/test/fp8e4m3fnuz.cpp#L35

File: test/onnx/verify/gemm_fp8_test.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    tmp = {0.10846, 0.672, 0.527, 0.94, 0.429, 0.2291};
+    std::vector<migraphx::fp8::fp8e4m3fnuz> c_data{tmp.cbegin(), tmp.cend()};
+
+    migraphx::parameter_map params;
+    params["A"] = migraphx::argument(a_shape, a_data.data());
+    params["B"] = migraphx::argument(b_shape, b_data.data());
+    params["C"] = migraphx::argument(c_shape, c_data.data());
+
+    auto result = p.eval(params).back();
+    std::vector<migraphx::fp8::fp8e4m3fnuz> result_vector;
```
Comment (umangyadav) [inline test/onnx/verify/gemm_fp8_test.cpp:63]: vector<fp8> worked ? If so use that for all the tests. In `add_fp8` test you are using `vector<float> for the gold values.

Comment (umangyadav) [inline test/onnx/verify/gemm_fp8_test.cpp:63]: vector<fp8> worked ? If so use that for all the tests. In `add_fp8` test you are using `vector<float> for the gold values.

Comment (umangyadav) [reply test/onnx/verify/gemm_fp8_test.cpp:63]: `verify_rms_range` would pick threshold based on data type of `results_vector`.

**Review by umangyadav — COMMENTED** [review]
As a next part we cover more operators for the "verification". 

e.g. "reduction" operations like reduce_max, reduce_sum, avgpool, maxpool 
math operations like sine, cosine, sqrt etc. 
binary ops : division and multiplication

File: test/onnx/verify/add_fp8_test.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    pp["0"] = l1.get_argument();
+    pp["1"] = l2.get_argument();
+
+    auto result = p.eval(pp).back();
+    std::vector<float> result_vector;
+    result.visit([&](auto output) { result_vector.assign(output.begin(), output.end()); });
+    std::vector<float> gold = {5.0};
```
Comment (CharlieL7) [inline test/onnx/verify/add_fp8_test.cpp:46]: As Umang mentioned, does `std::vector<migraphx::fp8::fp8e4m3fnuz>` work?

**Review by CharlieL7 — APPROVED** [review]
Address Umang's comment and should be good to go

File: test/onnx/verify/add_fp8_test.cpp (hunk 2 above)
Comment (ahsan-ca) [reply test/onnx/verify/add_fp8_test.cpp]: Done.

File: test/onnx/verify/gemm_fp8_test.cpp (hunk 3 above)
Comment (ahsan-ca) [reply test/onnx/verify/gemm_fp8_test.cpp:63]: Done.

Comment (ahsan-ca) [reply test/onnx/verify/gemm_fp8_test.cpp:63]: Yes, it worked. I have made the changes.

File: test/onnx/verify/add_fp8_test.cpp (hunk 4 above)
Comment (ahsan-ca) [reply test/onnx/verify/add_fp8_test.cpp]: Yes, it worked. I have made the changes.

File: test/onnx/parse/gemm_fp8_test.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    auto l2    = mm->add_parameter("C", migraphx::shape{migraphx::shape::fp8e4m3fnuz_type, {6, 1}});
+    auto alpha = 0.5f;
+    auto beta  = 0.8f;
+    auto a_l   = mm->add_literal(alpha);
+    auto t_a   = add_common_op(*mm, migraphx::make_op("mul"), {a_l, l0});
+    t_a        = mm->add_instruction(
+        migraphx::make_op("convert", {{"target_type", migraphx::shape::fp8e4m3fnuz_type}}), t_a);
+    t_a = mm->add_instruction(migraphx::make_op("transpose", {{"permutation", {1, 0}}}), t_a);
+    std::vector<std::size_t> lens = {6, 7};
+    auto dot = migraphx::add_apply_alpha_beta(*mm, {t_a, l1}, migraphx::make_op("dot"), 1.0f, 0.0f);
```
[the remark this answers is not in the document]
Comment (ahsan-ca) [reply test/onnx/parse/gemm_fp8_test.cpp:43]: Looking into this...

File: test/onnx/parse/shrink_fp8_test.cpp (hunk 1 above)
Comment (ahsan-ca) [reply test/onnx/parse/shrink_fp8_test.cpp:50]: Looking into this...

Comment (ahsan-ca) [reply test/onnx/parse/shrink_fp8_test.cpp:50]: Discussed this with Umang and need to update parse_shrink.cpp before making changes here. Opened PR: https://github.com/ROCm/AMDMIGraphX/pull/3142 to make the required change.

File: test/onnx/parse/gemm_fp8_test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahsan-ca) [reply test/onnx/parse/gemm_fp8_test.cpp]: Discussed this with Umang and removed `add_apply_alpha_beta`.

File: test/onnx/parse/gemm_fp8_test.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    std::vector<std::size_t> lens = {6, 7};
+    auto dot                      = mm->add_instruction(migraphx::make_op("dot"), t_a, l1);
+    l2 = mm->add_instruction(migraphx::make_op("multibroadcast", {{"out_lens", lens}}), l2);
+    l2 = mm->add_instruction(
+        migraphx::make_op("convert", {{"target_type", migraphx::shape::float_type}}), l2);
+    auto b_l  = mm->add_literal(beta);
+    auto b_b  = mm->add_instruction(migraphx::make_op("multibroadcast", {{"out_lens", lens}}), b_l);
+    auto l2_b = mm->add_instruction(migraphx::make_op("mul"), l2, b_b);
+    l2_b      = mm->add_instruction(
+        migraphx::make_op("convert", {{"target_type", migraphx::shape::fp8e4m3fnuz_type}}), l2_b);
```
Comment (umangyadav) [inline test/onnx/parse/gemm_fp8_test.cpp:52]: You can use `add_common_op` for this part.

File: test/onnx/verify/add_fp8_test.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    migraphx::parameter_map pp;
+    migraphx::literal l1{s, {3.25}};
+    migraphx::literal l2{s, {2.25}};
+    pp["0"] = l1.get_argument();
+    pp["1"] = l2.get_argument();
+
+    auto result = p.eval(pp).back();
+    std::vector<migraphx::fp8::fp8e4m3fnuz> result_vector;
+    result.visit([&](auto output) { result_vector.assign(output.begin(), output.end()); });
+    std::vector<migraphx::fp8::fp8e4m3fnuz> gold{static_cast<migraphx::fp8::fp8e4m3fnuz>(5.5)};
```
Comment (umangyadav) [inline test/onnx/verify/add_fp8_test.cpp:46]: Does it really require `static_cast` ?

File: test/onnx/verify/gemm_fp8_test.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    std::vector<float> tmp = {0.2646, 0.8525, 0.4192, 0.1415, 0.4321,  0.675,  0.4248, 0.8203,
+                              0.978,  0.5796, 0.6626, 0.479,  0.924,   0.734,  0.674,  0.8716,
+                              0.3733, 0.3328, 0.4272, 0.0247, 0.7583,  0.4873, 0.5835, 0.694,
+                              0.4375, 0.2406, 0.269,  0.6763, 0.542,   0.8994, 0.657,  0.5425,
+                              0.1412, 0.8994, 0.2183, 0.812,  0.937,   0.3438, 0.712,  0.9033,
+                              0.266,  0.8013, 0.803,  0.4993, 0.07196, 0.635,  0.7344, 0.3213};
+    std::vector<migraphx::fp8::fp8e4m3fnuz> a_data{tmp.cbegin(), tmp.cend()};
+
+    migraphx::shape b_shape{migraphx::shape::fp8e4m3fnuz_type, {8, 7}};
+    tmp = {0.7095,  0.612,  0.741,  0.02121, 0.3872, 0.4482,  0.6235,  0.02249, 0.2332, 0.7656,
```
Comment (umangyadav) [inline test/onnx/verify/gemm_fp8_test.cpp:45]: Can it be initialized directly without making use of `tmp` ?

File: test/onnx/verify/shrink_fp8_test.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+#include <migraphx/verify.hpp>
+#include <onnx_test.hpp>
+
+TEST_CASE(shrink_fp8_test)
+{
+    migraphx::program p = optimize_onnx("shrink_fp8_test.onnx");
+    p.compile(migraphx::make_target("ref"));
+
+    migraphx::shape s{migraphx::shape::fp8e4m3fnuz_type, {3, 3}};
+    // TODO: Make FP8 vector work for initializer list.
```
Comment (umangyadav) [inline test/onnx/verify/shrink_fp8_test.cpp:36]: Can you open an issue for this TODO and put all instances where `tmp` or` tmp_data` is used to instantiate fp8 vector ? using `tmp_data` and `tmp` vector doesn't look like a good idea. You can work on it in a different PR later.

File: test/onnx/verify/add_fp8_test.cpp (hunk 7 above)
Comment (ahsan-ca) [reply test/onnx/verify/add_fp8_test.cpp:46]: Without this, I get the following error:
```
/code/AMDMIGraphX/AMDMIGraphX/test/onnx/verify/add_fp8_test.cpp:46:50: error: type 'double' cannot be narrowed to 'size_type' (aka 'unsigned long') in initializer list [-Wc++11-narrowing]
    std::vector<migraphx::fp8::fp8e4m3fnuz> gold{5.5};
                                                 ^~~
/code/AMDMIGraphX/AMDMIGraphX/test/onnx/verify/add_fp8_test.cpp:46:50: note: insert an explicit cast to silence this issue
    std::vector<migraphx::fp8::fp8e4m3fnuz> gold{5.5};
                                                 ^~~
                                                 static_cast<size_type>( )
1 error generated.
```

File: test/onnx/verify/gemm_fp8_test.cpp (hunk 8 above)
Comment (ahsan-ca) [reply test/onnx/verify/gemm_fp8_test.cpp:45]: This is a workaround for the issue that FP8 initializer list is not working. I will open up an issue as suggested in the comment below to come to it later once we fix the initializer list issue.

File: test/onnx/verify/shrink_fp8_test.cpp (hunk 9 above)
Comment (ahsan-ca) [reply test/onnx/verify/shrink_fp8_test.cpp:36]: Sounds good. I will open up an issue for this.

**Review by umangyadav — APPROVED** [review]

**Pull request merged**